// @microsoft/teams.mcp/dist/plugin.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.McpPlugin = void 0;
const mcp_js_1 = require("@modelcontextprotocol/sdk/server/mcp.js");
const sse_js_1 = require("@modelcontextprotocol/sdk/server/sse.js");
const stdio_js_1 = require("@modelcontextprotocol/sdk/server/stdio.js");
const json_schema_to_zod_1 = require("json-schema-to-zod");
const teams_apps_1 = require("@microsoft/teams.apps");
const teams_dev_1 = require("@microsoft/teams.dev");
const package_json_1 = __importDefault(require("../package.json"));
/**
 * High-level MCP server that provides a simpler API for working with resources, tools, and prompts.
 * For advanced usage (like sending notifications or setting custom request handlers),
 * use the underlying Server instance available via the server property.
 */
let McpPlugin = class McpPlugin {
    logger;
    httpPlugin;
    devtoolsPlugin;
    server;
    id = -1;
    inspector;
    connections = {};
    transport = {
        type: 'sse',
    };
    constructor(options = {}) {
        this.inspector =
            options instanceof mcp_js_1.McpServer
                ? 'http://localhost:5173'
                : options.inspector || 'http://localhost:5173';
        this.server =
            options instanceof mcp_js_1.McpServer
                ? options
                : new mcp_js_1.McpServer({
                    name: options.name || 'mcp',
                    version: options.version || '0.0.0',
                }, options);
        if (!(options instanceof mcp_js_1.McpServer) && options.transport) {
            this.transport = options.transport;
        }
    }
    /**
     * add a chat prompt to your server
     * @param prompt the chat prompt
     */
    use(prompt) {
        for (const fn of prompt.functions) {
            const schema = eval((0, json_schema_to_zod_1.jsonSchemaToZod)(fn.parameters, { module: 'cjs' }));
            this.server.tool(fn.name, fn.description, schema.shape, this.onToolCall(fn.name, prompt));
        }
        return this;
    }
    /**
     * Pass through call to the underlying MCP server
     */
    tool(...params) {
        this.server.tool(...params);
        return this;
    }
    /**
     * Pass through call to the underlying MCP server
     */
    prompt(...params) {
        this.server.prompt(...params);
        return this;
    }
    /**
     * Pass through call to the underlying MCP server
     */
    resource(...params) {
        this.server.resource(...params);
        return this;
    }
    onInit() {
        this.devtoolsPlugin?.addPage({
            name: 'mcp',
            displayName: 'MCP',
            url: this.inspector,
        });
        if (this.transport.type === 'sse') {
            return this.onInitSSE(this.httpPlugin, this.transport);
        }
        return this.onInitStdio(this.transport);
    }
    onStart({ port }) {
        if (this.transport.type === 'sse') {
            this.logger.info(`listening at http://localhost:${port}${this.transport.path || '/mcp'}`);
        }
        else {
            this.logger.info('listening on stdin');
        }
    }
    onInitStdio(options) {
        const transport = new stdio_js_1.StdioServerTransport(options.stdin, options.stdout);
        return this.server.connect(transport);
    }
    onInitSSE(http, options) {
        const path = options.path || '/mcp';
        http.get(path, (_, res) => {
            this.id++;
            this.logger.debug('connecting...');
            const transport = new sse_js_1.SSEServerTransport(`${path}/${this.id}/messages`, res);
            this.connections[this.id] = {
                id: this.id,
                transport,
                createdAt: new Date(),
            };
            this.server.connect(transport);
        });
        http.post(`${path}/:id/messages`, (req, res) => {
            const id = +req.params.id;
            const { transport } = this.connections[id];
            if (!transport) {
                res.status(401).send('unauthorized');
                return;
            }
            transport.handlePostMessage(req, res);
        });
    }
    onToolCall(name, prompt) {
        return async (args) => {
            try {
                const res = await prompt.call(name, args);
                if (this.isCallToolResult(res)) {
                    return res;
                }
                return {
                    content: [
                        {
                            type: 'text',
                            text: typeof res === 'string' ? res : JSON.stringify(res),
                        },
                    ],
                };
            }
            catch (err) {
                this.logger.error(err.toString());
                return {
                    isError: true,
                    content: [
                        {
                            type: 'text',
                            text: err.toString(),
                        },
                    ],
                };
            }
        };
    }
    isCallToolResult(value) {
        if (!!value || !('content' in value))
            return false;
        const { content } = value;
        return (Array.isArray(content) &&
            content.every((item) => 'type' in item &&
                (item.type === 'text' ||
                    item.type === 'image' ||
                    item.type === 'resource')));
    }
};
exports.McpPlugin = McpPlugin;
__decorate([
    (0, teams_apps_1.Logger)(),
    __metadata("design:type", Object)
], McpPlugin.prototype, "logger", void 0);
__decorate([
    (0, teams_apps_1.Dependency)(),
    __metadata("design:type", teams_apps_1.HttpPlugin)
], McpPlugin.prototype, "httpPlugin", void 0);
__decorate([
    (0, teams_apps_1.Dependency)({ optional: true }),
    __metadata("design:type", teams_dev_1.DevtoolsPlugin)
], McpPlugin.prototype, "devtoolsPlugin", void 0);
exports.McpPlugin = McpPlugin = __decorate([
    (0, teams_apps_1.Plugin)({
        name: 'mcp',
        version: package_json_1.default.version,
        description: [
            'High-level MCP server that provides a simpler API for working with resources, tools, and prompts.',
            'For advanced usage (like sending notifications or setting custom request handlers),',
            'use the underlying Server instance available via the server property.',
        ].join('\n'),
    }),
    __metadata("design:paramtypes", [Object])
], McpPlugin);
